### PR TITLE
codec_adapter: fix deep_buffer_bytes calculation

### PR DESCRIPTION
**Pull request by ranj063**
Fix the deep_buffer_bytes calculation to use the out_buff_size instead
of in_buff_size. Remove the default periods for the local buffer as the
calculation should work out to 2 when period bytes and out_buff_size are
equal.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

### CODE REVIEW

**Comment by ranj063** [comment]
@dbaluta let me know what you think of this. Ideally I'd think that deep_buffer_bytes should also be based on out_buff_size but it completely break all the cadence codec tests in CI

**Review by lgirdwood — APPROVED** [review]

File: src/audio/codec_adapter/codec_adapter.c [hunk 1]
```diff
@@ -166,8 +166,19 @@ int codec_adapter_prepare(struct comp_dev *dev)
...
+		buff_periods = (mod->period_bytes % md->mpd.out_buff_size) ?
+			       (mod->period_bytes / md->mpd.out_buff_size) + 2 :
+			       (mod->period_bytes / md->mpd.out_buff_size) + 1;
+	}
+
 	buff_size = MAX(mod->period_bytes, md->mpd.out_buff_size) * buff_periods;
+
+	/* Allocate local buffer */
 	if (mod->local_buff) {
 		ret = buffer_set_size(mod->local_buff, buff_size);
```
Comment (dbaluta) [inline src/audio/codec_adapter/codec_adapter.c:183]: so, `deep_buff_bytes` depends on the  `in_buff_size` and `period_bytes`, while local_buff_size depends on `period_bytes` and `out_buff_size`. 

Makes sense.

**Review by dbaluta — APPROVED** [review]

**Comment by dbaluta** [comment]
@ranj063 we need to draw some conclusion with this and maybe add them in the commit message as it might be useful for people understanding the whole processing behing codec_adapter.

i tested this with mp3 and it works fine.

So, we need to answer the following question:

Q:
- What is deep_buffer_bytes used for?
A:
- I think deep_buffer_bytes is a measure of how many bytes do we need to send to the DAI before the codec adapter starts producing something.
In a normal copy() walk it might be possible that the first period_bytes copied to input_buffer might not be enough for the processing to take place. Anyhow, in order not to starve the DAI we need to feed it something. That would be the generate_zeroes() function.

Q:
- What is local_buffer used for?
A:
- Sometimes / Usually the codec_adapter process function will produce more data then period_bytes (DAI consume period_bytes). So, in order not to lose the bytes we copy them to the local_buffer and the DAI will copy them from local_buffer.


Now, there is also a flow control mechanism in place. Because usually we produce a lot more than we consume at some point there will be no more space in codec_adapter input_buffer so the Host will run doing nothing (I usually see the message "no bytes to copy in host).
Now, this is still fine but a little bit inefficient. Once the codec_adapter feeds some more data to the DAI, the input buffer of codec_adapter will be emptied and again Host will be able to send next period_bytes.

An optimization for this would be to have different 'period_bytes' for Host and for DAI. But would mean to split it in two separate pipelines.

File: src/audio/codec_adapter/codec_adapter.c [hunk 2]
```diff
@@ -173,8 +173,24 @@ int codec_adapter_prepare(struct comp_dev *dev)
...
 
-	/* Allocate local buffer */
+	if (md->mpd.out_buff_size > mod->period_bytes) {
+		buff_periods = (md->mpd.out_buff_size % mod->period_bytes) ?
+			       (md->mpd.out_buff_size / mod->period_bytes) + 2 :
+			       (md->mpd.out_buff_size / mod->period_bytes) + 1;
+	} else {
+		buff_periods = (mod->period_bytes % md->mpd.out_buff_size) ?
+			       (mod->period_bytes / md->mpd.out_buff_size) + 2 :
+			       (mod->period_bytes / md->mpd.out_buff_size) + 1;
```
Comment (kv2019i) [inline src/audio/codec_adapter/codec_adapter.c:183]: Some readability notes. This same logic is repeating four time in this function, so a macro or inlined helper function would be helpful. Also reusing "buff_periods" for different purposes is a bit distracting. 

But alas, neither is added with this patch, so not really a blocker.

**Review by kv2019i — APPROVED** [review]
Some nitpicks on code readability, but nothing really added by this patch, and the fixed logic seems correct.

**Review by lgirdwood — APPROVED** [review]

File: src/audio/codec_adapter/codec_adapter.c (hunk 2 above)
Comment (lgirdwood) [reply src/audio/codec_adapter/codec_adapter.c:183]: Lets cleanup the readability in a subsequent update, I know @ranj063 has more in progress

**Pull request merged**